RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplateReco2D.cc

```diff
@@ -0,0 +1,654 @@
+//! \param       qbin - (output) index (0-4) describing the charge of the cluster
+//!                       qbin = 0        Q/Q_avg > 1.5   [few % of all hits]
+//!                              1  1.5 > Q/Q_avg > 1.0   [~30% of all hits]
+//!                              2  1.0 > Q/Q_avg > 0.85  [~30% of all hits]
+//!                              3 0.85 > Q/Q_avg > min1  [~30% of all hits]
+//! \param     deltay - (output) template y-length - cluster length [when > 0, possibly missing end]
+// *************************************************************************************************************************************
+int SiPixelTemplateReco2D::PixelTempReco3D(int id, float cotalpha, float cotbeta, float locBz, float locBx, int edgeflagy, int edgeflagx,
+                                           ClusMatrix & cluster, SiPixelTemplate2D& templ2D,float& yrec, float& sigmay,
+                                           float& xrec, float& sigmax, float& probxy, float& probQ, int& qbin, float& deltay, int& npixels)
+
+{
+   // Local variables
+   int i, j, k;
+   float template2d[BXM2][BYM2], dpdx2d[2][BXM2][BYM2], fbin[3];
+   
+   // fraction of truncation signal to measure the cluster ends
+   const float fracpix = 0.45f;
+   
+   
+   // Extract some relevant info from the 2D template
+   
+   if(id > 0) {templ2D.interpolate(id, cotalpha, cotbeta, locBz, locBx);}
+   float xsize = templ2D.xsize();
+   float ysize = templ2D.ysize();
+   
+   // Allow Qbin Q/Q_avg fractions to vary to optimize error estimation
+   
+   for(i=0; i<3; ++i) {fbin[i] = templ2D.fbin(i);}
+   
+   float q50 = templ2D.s50();
+   float pseudopix = 0.1f*q50;
+   float pseudopix2 = q50*q50;
+   
+   
+   // Get charge scaling factor
+   
+   float qscale = templ2D.qscale();
+   
+   // Check that the cluster container is (up to) a 7x21 matrix and matches the dimensions of the double pixel flags
+   
+   int nclusx = cluster.mrow;
+   int nclusy = (int)cluster.mcol;
+   bool * xdouble = cluster.xdouble;
+   bool * ydouble = cluster.ydouble;
+   
+   // First, rescale all pixel charges and compute total charge
+   float qtotal = 0.f;
+   for(i=0; i<nclusx*nclusy; ++i) {
+      cluster.matrix[i] *= qscale;
+      qtotal +=cluster.matrix[i];
+   }
+   
+   // uncertainty and final corrections depend upon total charge bin
+   
+   float fq0 = qtotal/templ2D.qavg();
+   
+   // Next, copy the cluster and double pixel flags into a shifted workspace to
+   // allow for significant zeros [pseudopixels] to be added
+   
+   float clusxy[BXM2][BYM2];
+   for(j=0; j<BXM2; ++j) {for(i=0; i<BYM2; ++i) {clusxy[j][i] = 0.f;}}
+   
+   int indexxy[2][NPIXMAX];
+   float pixel[NPIXMAX];
+   float sigma2[NPIXMAX];
+   float minmax = templ2D.pixmax();
+   int imin=BYM2, imax=0, jmin=BXM2, jmax=0;
+   float ylow0 = 0.f, yhigh0 = 0.f, xlow0 = 0.f, xhigh0 = 0.f;
+   float ye = 0.f, xe;
+   int npixel = 0;
+   float ysum[BYM2], ypos[BYM2];
+   //   float xq = 0.f, qtot = 0.f;

```

Comment:
If they are not needed, just remove them

Reply:
I don't know.  I'll need to discuss with Morris.  I propose to postpone this until later.